src/fsharp/PrettyNaming.fs

```diff
@@ -109,66 +109,162 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
         ( ')', "RParen");
         ( ' ', "Space");
         ( '[', "LBrack");
-        ( ']', "RBrack"); ]
+        ( ']', "RBrack"); |]
 
-    let private opCharDict = 
-        let t = new Dictionary<_,_>()
-        for (c,_) in opCharTranslateTable do 
-            t.Add(c,1)
+    /// The set of characters usable in custom operators.
+    let private opCharSet =
+        let t = new HashSet<_>()
+        for (c,_) in opCharTranslateTable do
+            t.Add(c) |> ignore
         t
         
-    let IsOpName (n:string) =
-        let rec loop i = (i < n.Length && (opCharDict.ContainsKey(n.[i]) || loop (i+1)))
+    let IsOpName (name:string) =
+        let nameLen = name.Length
+        let rec loop i = (i < nameLen && (opCharSet.Contains(name.[i]) || loop (i+1)))
         loop 0
 
+    let IsMangledOpName (n:string) =
+        n.StartsWith (opNamePrefix, System.StringComparison.Ordinal)
+
+    // +++ GLOBAL STATE
+    /// Compiles a custom operator into a mangled operator name.
+    /// For example, "!%" becomes "op_DereferencePercent"
+    let private compileCustomOpName =
+        let t2 =
+            let t2 = Dictionary<_,_> (opCharTranslateTable.Length)
+            for x, y in opCharTranslateTable do
+                t2.Add (x, y)
+            t2
```

Comment:
Not sure if the `dict` function is defined/available here, but if it is, this would simplify to just `let t2 = dict opCharTranslateTable`